File: src/Lexer.java
// Y++ Lexer — converts raw source into a list of Tokens
package ypp;

import java.util.ArrayList;
import java.util.List;

public class Lexer {

    private final String source;
    private int pos  = 0;
    private int line = 1;

    public Lexer(String source) {
        // Strip block comments  \\...\\  before tokenising
        this.source = stripComments(source);
    }

    // ---------------------------------------------------------------
    // Public API
    // ---------------------------------------------------------------

    public List<Token> tokenize() {
        List<Token> tokens = new ArrayList<>();
        while (pos < source.length()) {
            skipWhitespace();
            if (pos >= source.length()) break;

            Token t = nextToken();
            if (t != null) tokens.add(t);
        }
        tokens.add(new Token(Token.Type.EOF, "", line));
        return tokens;
    }

    // ---------------------------------------------------------------
    // Comment stripping  \\...\\
    // ---------------------------------------------------------------

    private static String stripComments(String src) {
        StringBuilder sb = new StringBuilder();
        int i = 0;
        while (i < src.length()) {
            if (i + 1 < src.length() && src.charAt(i) == '\\' && src.charAt(i + 1) == '\\') {
                // Find closing \\
                int end = src.indexOf("\\\\", i + 2);
                if (end == -1) {
                    // Unclosed comment — skip to end
                    break;
                }
                // Replace comment text with spaces to preserve line numbers
                for (int j = i; j < end + 2; j++) {
                    sb.append(src.charAt(j) == '\n' ? '\n' : ' ');
                }
                i = end + 2;
            } else {
                sb.append(src.charAt(i));
                i++;
            }
        }
        return sb.toString();
    }

    // ---------------------------------------------------------------
    // Core tokeniser
    // ---------------------------------------------------------------

    private void skipWhitespace() {
        while (pos < source.length()) {
            char c = source.charAt(pos);
            if (c == '\n') { line++; pos++; }
            else if (Character.isWhitespace(c)) pos++;
            else break;
        }
    }

    private Token nextToken() {
        int startLine = line;
        char c = source.charAt(pos);

        // String literal
        if (c == '"') return readString(startLine);

        // Number literal (may carry si / i / d suffix)
        if (Character.isDigit(c) || (c == '-' && pos + 1 < source.length() && Character.isDigit(source.charAt(pos + 1)))) {
            return readNumber(startLine);
        }

        // Identifier / keyword
        if (Character.isLetter(c) || c == '_') return readIdent(startLine);

        // Single/double character symbols
        pos++;
        return switch (c) {
            case '=' -> new Token(Token.Type.EQUALS,    "=", startLine);
            case '*' -> new Token(Token.Type.STAR,      "*", startLine);
            case '+' -> new Token(Token.Type.PLUS,      "+", startLine);
            case '-' -> new Token(Token.Type.MINUS,     "-", startLine);
            case '/' -> new Token(Token.Type.SLASH,     "/", startLine);
            case '.' -> new Token(Token.Type.DOT,       ".", startLine);
            case '!' -> new Token(Token.Type.BANG,      "!", startLine);
            case ':' -> {
                if (pos < source.length() && source.charAt(pos) == ':') {
                    pos++;
                    yield new Token(Token.Type.DOUBLE_COLON, "::", startLine);
                }
                yield new Token(Token.Type.COLON, ":", startLine);
            }
            case ';' -> new Token(Token.Type.SEMICOLON, ";", startLine);
            case ',' -> new Token(Token.Type.COMMA,     ",", startLine);
            case '{' -> new Token(Token.Type.LBRACE,    "{", startLine);
            case '}' -> new Token(Token.Type.RBRACE,    "}", startLine);
            case '(' -> new Token(Token.Type.LPAREN,    "(", startLine);
            case ')' -> new Token(Token.Type.RPAREN,    ")", startLine);
            default  -> null;   // skip unrecognised characters silently
        };
    }

    // ---------------------------------------------------------------
    // String  "..."
    // ---------------------------------------------------------------

    private Token readString(int startLine) {
        pos++; // skip opening "
        StringBuilder sb = new StringBuilder();
        while (pos < source.length() && source.charAt(pos) != '"') {
            if (source.charAt(pos) == '\n') line++;
            sb.append(source.charAt(pos++));
        }
        if (pos < source.length()) pos++; // skip closing "
        return new Token(Token.Type.LIT_STRING, sb.toString(), startLine);
    }

    // ---------------------------------------------------------------
    // Numbers:  2.4d   3i   1si   (bare numbers used as labels)
    // ---------------------------------------------------------------

    private Token readNumber(int startLine) {
        StringBuilder sb = new StringBuilder();
        // Optional leading minus
        if (source.charAt(pos) == '-') sb.append(source.charAt(pos++));

        while (pos < source.length() && (Character.isDigit(source.charAt(pos)) || source.charAt(pos) == '.')) {
            sb.append(source.charAt(pos++));
        }

        // Read optional suffix: si / i / d
        String suffix = readSuffix();
        String numText = sb.toString();

        return switch (suffix) {
            case "si" -> new Token(Token.Type.LIT_SMALLINT, numText, startLine);
            case "i"  -> new Token(Token.Type.LIT_INTEGER,  numText, startLine);
            case "d"  -> new Token(Token.Type.LIT_DOUBLE,   numText, startLine);
            default   -> new Token(Token.Type.NUMBER_LABEL, numText, startLine);
        };
    }

    /** Peek ahead and consume  si / i / d  suffix if present. */
    private String readSuffix() {
        if (pos >= source.length()) return "";
        // "si" must come before "i" check
        if (pos + 1 < source.length()
                && source.charAt(pos) == 's'
                && source.charAt(pos + 1) == 'i') {
            pos += 2;
            return "si";
        }
        char c = source.charAt(pos);
        if (c == 'i' || c == 'd') {
            pos++;
            return String.valueOf(c);
        }
        return "";
    }

    // ---------------------------------------------------------------
    // Identifiers & Keywords
    // ---------------------------------------------------------------

    private Token readIdent(int startLine) {
        StringBuilder sb = new StringBuilder();
        while (pos < source.length() && (Character.isLetterOrDigit(source.charAt(pos)) || source.charAt(pos) == '_')) {
            sb.append(source.charAt(pos++));
        }
        String word = sb.toString();

        // Check for cast calls: int(), double(), string(), stringint()
        // They appear as bare words immediately followed by ()
        if ((word.equals("int") || word.equals("double") || word.equals("string") || word.equals("stringint")) && peekTwoChars("()")) {
            pos += 2; // consume ()
            Token.Type ctype = switch (word) {
                case "int" -> Token.Type.CAST_INT;
                case "double" -> Token.Type.CAST_DOUBLE;
                case "string" -> Token.Type.CAST_STRING;
                default -> Token.Type.CAST_STRINGINT;
            };
            return new Token(ctype, word + "()", startLine);
        }

        return switch (word) {
            case "Import"    -> new Token(Token.Type.IMPORT,       word, startLine);
            case "PRINT"     -> new Token(Token.Type.PRINT,        word, startLine);
            case "NUM"       -> new Token(Token.Type.NUM,          word, startLine);
            case "STRING"    -> new Token(Token.Type.STRING_BLOCK, word, startLine);
            case "EXCEPTION" -> new Token(Token.Type.EXCEPTION,    word, startLine);
            case "CONCAT"    -> new Token(Token.Type.CONCAT,       word, startLine);
            case "Public"    -> new Token(Token.Type.PUBLIC,       word, startLine);
            case "class"     -> new Token(Token.Type.CLASS,        word, startLine);
            case "func"      -> new Token(Token.Type.FUNC,         word, startLine);
            case "NEW", "new"-> new Token(Token.Type.NEW,          word, startLine);
            case "global"    -> new Token(Token.Type.KW_GLOBAL,    word, startLine);
            case "while"     -> new Token(Token.Type.WHILE,        word, startLine);
            case "NOT"       -> new Token(Token.Type.NOT,          word, startLine);
            case "smallint"  -> new Token(Token.Type.KW_SMALLINT,  word, startLine);
            case "integer"   -> new Token(Token.Type.KW_INTEGER,   word, startLine);
            case "double"    -> new Token(Token.Type.KW_DOUBLE,    word, startLine);
            case "slong"     -> new Token(Token.Type.KW_SLONG,     word, startLine);
            case "schar"     -> new Token(Token.Type.KW_SCHAR,     word, startLine);
            default          -> new Token(Token.Type.IDENT,        word, startLine);
        };
    }

    private boolean peekTwoChars(String s) {
        return pos + 1 < source.length()
            && source.charAt(pos) == s.charAt(0)
            && source.charAt(pos + 1) == s.charAt(1);
    }
}

File: lexer.jl
# Y++ Lexer - Tokenizes Y++ source code

@enum TokenType begin
    # Keywords
    NVAR
    SVAR
    PRINT
    FUNC
    
    # Literals
    NUMBER
    STRING
    IDENTIFIER
    
    # Operators and Punctuation
    EQUALS
    SEMICOLON
    COMMA
    LPAREN
    RPAREN
    LBRACE
    RBRACE
    LBRACKET
    RBRACKET
    DOT
    
    # Special
    EOF
    NEWLINE
end

struct Token
    type::TokenType
    value::Any
    line::Int
    column::Int
end

mutable struct Lexer
    input::String
    position::Int
    line::Int
    column::Int
    current_char::Union{Char, Nothing}
    
    function Lexer(input::String)
        lexer = new(input, 1, 1, 1, nothing)
        if length(input) > 0
            lexer.current_char = input[1]
        end
        return lexer
    end
end

function advance!(lexer::Lexer)
    if lexer.position >= length(lexer.input)
        lexer.current_char = nothing
        return
    end
    
    if lexer.current_char == '\n'
        lexer.line += 1
        lexer.column = 1
    else
        lexer.column += 1
    end
    
    lexer.position += 1
    if lexer.position <= length(lexer.input)
        lexer.current_char = lexer.input[lexer.position]
    else
        lexer.current_char = nothing
    end
end

function peek(lexer::Lexer, offset::Int=1)::Union{Char, Nothing}
    peek_pos = lexer.position + offset
    if peek_pos <= length(lexer.input)
        return lexer.input[peek_pos]
    end
    return nothing
end

function skip_whitespace!(lexer::Lexer)
    while lexer.current_char !== nothing && lexer.current_char in [' ', '\t', '\r']
        advance!(lexer)
    end
end

function read_number(lexer::Lexer)::Float64
    num_str = ""
    while lexer.current_char !== nothing && (isdigit(lexer.current_char) || lexer.current_char == '.')
        num_str *= lexer.current_char
        advance!(lexer)
    end
    return parse(Float64, num_str)
end

function read_string(lexer::Lexer)::String
    str = ""
    advance!(lexer)  # Skip opening quote
    
    while lexer.current_char !== nothing && lexer.current_char != '"'
        str *= lexer.current_char
        advance!(lexer)
    end
    
    if lexer.current_char == '"'
        advance!(lexer)  # Skip closing quote
    end
    
    return str
end

function read_identifier(lexer::Lexer)::String
    id = ""
    while lexer.current_char !== nothing && (isalnum(lexer.current_char) || lexer.current_char == '_')
        id *= lexer.current_char
        advance!(lexer)
    end
    return id
end

function get_next_token(lexer::Lexer)::Token
    while lexer.current_char !== nothing
        # Skip whitespace
        if lexer.current_char in [' ', '\t', '\r']
            skip_whitespace!(lexer)
            continue
        end
        
        # Handle newlines
        if lexer.current_char == '\n'
            line, col = lexer.line, lexer.column
            advance!(lexer)
            return Token(NEWLINE, "\n", line, col)
        end
        
        # Handle numbers
        if isdigit(lexer.current_char)
            line, col = lexer.line, lexer.column
            return Token(NUMBER, read_number(lexer), line, col)
        end
        
        # Handle strings
        if lexer.current_char == '"'
            line, col = lexer.line, lexer.column
            return Token(STRING, read_string(lexer), line, col)
        end
        
        # Handle identifiers and keywords
        if isalpha(lexer.current_char) || lexer.current_char == '_'
            line, col = lexer.line, lexer.column
            id = read_identifier(lexer)
            
            # Check for keywords
            token_type = if id == "nvar"
                NVAR
            elseif id == "svar"
                SVAR
            elseif id == "PRINT"
                PRINT
            elseif id == "Func"
                FUNC
            else
                IDENTIFIER
            end
            
            return Token(token_type, id, line, col)
        end
        
        # Single character tokens
        line, col = lexer.line, lexer.column
        char = lexer.current_char
        
        if char == '='
            advance!(lexer)
            return Token(EQUALS, "=", line, col)
        elseif char == ';'
            advance!(lexer)
            return Token(SEMICOLON, ";", line, col)
        elseif char == ','
            advance!(lexer)
            return Token(COMMA, ",", line, col)
        elseif char == '('
            advance!(lexer)
            return Token(LPAREN, "(", line, col)
        elseif char == ')'
            advance!(lexer)
            return Token(RPAREN, ")", line, col)
        elseif char == '{'
            advance!(lexer)
            return Token(LBRACE, "{", line, col)
        elseif char == '}'
            advance!(lexer)
            return Token(RBRACE, "}", line, col)
        elseif char == '['
            advance!(lexer)
            return Token(LBRACKET, "[", line, col)
        elseif char == ']'
            advance!(lexer)
            return Token(RBRACKET, "]", line, col)
        elseif char == '.'
            advance!(lexer)
            return Token(DOT, ".", line, col)
        elseif char == ':'
            advance!(lexer)
            continue  # Skip colons
        else
            error("Unexpected character: '$char' at line $(lexer.line), column $(lexer.column)")
        end
    end
    
    return Token(EOF, nothing, lexer.line, lexer.column)
end

function tokenize(input::String)::Vector{Token}
    lexer = Lexer(input)
    tokens = Token[]
    
    while true
        token = get_next_token(lexer)
        if token.type != NEWLINE  # Skip newline tokens for simplicity
            push!(tokens, token)
        end
        if token.type == EOF
            break
        end
    end
    
    return tokens
end

# Made with Bob
